### PR TITLE
Add GEP to change DNS balancer to a software loadbalancer

### PR DESCRIPTION
**Pull request by whiterm**
Change DNS balancer to a software load balancer

### CODE REVIEW

File: enhancements/gep/001-loadbalancer/README.md [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+## Summary
+The main goal of this proposal is to replace dns balancing with a software loadbalancer.
+Thus, all requests to the kube-apiserver from Kubernetes components (kubelet, scheduler, control-manager, kube-proxy)
+will go through this loadbalancer. Also, requests to the docker registry must also go through the loadbalancer.
+
+## Motivation
+At the moment, Gravity is using CoreDNS to determine which master to send requests to,
+but DNS is not a good tool for balancing traffic.
+Implementation of this proposal will provide the following improvements
```
Comment (a-palchikov) [inline enhancements/gep/001-loadbalancer/README.md]:
```suggestion
This proposal aims to provide the following improvements:
```

File: enhancements/gep/001-loadbalancer/README.md [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+  1. will evenly distribute the load on the apiserver kubernetes
```
Comment (a-palchikov) [inline enhancements/gep/001-loadbalancer/README.md]:
```suggestion
  1. Evenly distribute the load between all apiserver replicas
```

File: enhancements/gep/001-loadbalancer/README.md [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+  1. Improve the stability and scalability of the cluster
+  1. nodes will react faster on the disaster of the masters and recovery process
```
Comment (a-palchikov) [inline enhancements/gep/001-loadbalancer/README.md]:
```suggestion
  1. Nodes will react faster to master failures [and recovery process]
```
not sure I understand the bit about the recovery process. Is this intending to say that the recovery will also be snappier?

File: enhancements/gep/001-loadbalancer/README.md [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+  1. the update will be faster and more efficient
```
Comment (a-palchikov) [inline enhancements/gep/001-loadbalancer/README.md]:
```suggestion
  1. Cluster upgrades will be faster and more efficient
```
Putting a distinction between an update (minor change aka patch) and upgrade (renewing software to the next version).

File: enhancements/gep/001-loadbalancer/README.md [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+
+### Goals
+When creating a cluster, the user can choose which load balancer will be external or internal.
```
Comment (a-palchikov) [inline enhancements/gep/001-loadbalancer/README.md]:
```suggestion
When creating a cluster, the user can choose the load balancer to be either external or internal.
```

File: enhancements/gep/001-loadbalancer/README.md [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+External loadbalancer is not controlled or configured by Gravity, it is the user's responsibility.
+The internal loadbalancer, on the contrary, is controlled and configured by Gravity, the planet-agent is responsible for this.
+When using an internal loadbalancer, a loadbalancer (HAProxy) will be installed for each node in the cluster including the master.
```
Comment (a-palchikov) [inline enhancements/gep/001-loadbalancer/README.md]:
```suggestion
When using an internal loadbalancer, a loadbalancer (HAProxy) will be installed for each node in the cluster including the master nodes.
```

File: enhancements/gep/001-loadbalancer/README.md [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+When changing the configuration of the masters, the agent should automatically change the loadbalancer configuration on each node.
```
Comment (a-palchikov) [inline enhancements/gep/001-loadbalancer/README.md]: Which configuration do you mean? As in upgrade generating the new planet container configuration or anything else?

File: enhancements/gep/001-loadbalancer/README.md [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+All Kubernetes components (kubelet, scheduler, control-manager, kube-proxy) will use the loadbalancer to access the apiserver.
```
Comment (a-palchikov) [inline enhancements/gep/001-loadbalancer/README.md:44]: Aren't they already using the loadbalancing nature of the kubernetes services? The question is why we're switching one balancer to the other as it has not been explained yet?

File: enhancements/gep/001-loadbalancer/README.md [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+
+### Non-Goals
+
+## Proposal
+Change DNS balancer to a software loadbalancer (HAProxy).
+CoreDNS will resolve the `leader.telekube.local` address, which will point either to the external address (domain name or IP) of the loadbalancer or to 127.0.0.1 when using the internal loadbalancer address.
```
Comment (a-palchikov) [inline enhancements/gep/001-loadbalancer/README.md]:
```suggestion
CoreDNS will resolve `leader.telekube.local` to either the external address (domain name or IP) of the loadbalancer (subject to which configuration?) or to `127.0.0.1` when using the internal loadbalancer.
```

File: enhancements/gep/001-loadbalancer/README.md [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+Image names should remain unchanged and start with `leader.telekube.local:5000`.
```
Comment (a-palchikov) [inline enhancements/gep/001-loadbalancer/README.md]:
```suggestion
Docker image references should continue pointing to internal registry at `leader.telekube.local:5000`.
```

File: enhancements/gep/001-loadbalancer/README.md [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+To do this, the docker register must use port 5001, and port 5000 must be open on the loadbalancer.
```
Comment (a-palchikov) [inline enhancements/gep/001-loadbalancer/README.md]:
```suggestion
To achieve this, docker registry will use port 5001 with port 5000 left to the loadbalancer.
```

File: enhancements/gep/001-loadbalancer/README.md [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+
+### User Stories (Optional)
```
Comment (a-palchikov) [inline enhancements/gep/001-loadbalancer/README.md]: Why optional - this can potentially implemented as integration tests.

File: enhancements/gep/001-loadbalancer/README.md [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+#### Story 1
+As a cluster administrator, I can set the loadbalancer settings when creating a cluster. Set the type of loadbalancer external or internal and indicate the external address.
```
Comment (a-palchikov) [inline enhancements/gep/001-loadbalancer/README.md]:
```suggestion
As a cluster administrator, I can configure the loadbalancer when creating a cluster. If I set the type of loadbalancer to external, I can provide the external address in configuration.
```

File: enhancements/gep/001-loadbalancer/README.md [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+#### Story 2
+As a cluster administrator, I can change the loadbalancer settings on a running cluster
```
Comment (a-palchikov) [inline enhancements/gep/001-loadbalancer/README.md]:
```suggestion
As a cluster administrator, I can change the loadbalancer settings on a running cluster.
```

File: enhancements/gep/001-loadbalancer/README.md [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+
+### Risks and Mitigations
+The `leader.telekube.local` address will not point to the master node.
```
Comment (a-palchikov) [inline enhancements/gep/001-loadbalancer/README.md]:
```suggestion
As the `leader.telekube.local` domain will not point to a master node, the applications using this address should be updated accordingly.
```
This is a bit confusing as gravity is the biggest application using it. Also, this provides no hints for the applications still using the domain - what should change for them?

File: enhancements/gep/001-loadbalancer/README.md [hunk 16]
```diff
@@ -0,0 +1,118 @@
...
+kind: ClusterConfiguration
+version: v1
+spec:
+  loadbalancer:
+    # default value is internal
+    type: "external|internal"
+    # gravity uses this field when type is external
+    externalAddress: "IP address | DNS address"
+```
+The names of docker images should not be changed, and they should start with `leader.telekube.local:5000/`.
```
Comment (a-palchikov) [inline enhancements/gep/001-loadbalancer/README.md]:
```suggestion
The docker image references should not be changed, and they should start with `leader.telekube.local:5000/`. 
```

File: enhancements/gep/001-loadbalancer/README.md [hunk 17]
```diff
@@ -0,0 +1,118 @@
...
+version: v1
+spec:
+  loadbalancer:
+    # default value is internal
+    type: "external|internal"
+    # gravity uses this field when type is external
+    externalAddress: "IP address | DNS address"
+```
+The names of docker images should not be changed, and they should start with `leader.telekube.local:5000/`. 
+So docker registers on master nodes by default will use port 5001 instead of 5000,
```
Comment (a-palchikov) [inline enhancements/gep/001-loadbalancer/README.md]:
```suggestion
So docker registries on master nodes by default will use port 5001 with 5000 reserved for the loadbalancer. 
```

File: enhancements/gep/001-loadbalancer/README.md [hunk 18]
```diff
@@ -0,0 +1,118 @@
...
+spec:
+  loadbalancer:
+    # default value is internal
+    type: "external|internal"
+    # gravity uses this field when type is external
+    externalAddress: "IP address | DNS address"
+```
+The names of docker images should not be changed, and they should start with `leader.telekube.local:5000/`. 
+So docker registers on master nodes by default will use port 5001 instead of 5000, 
+and loadbalancer will use port 5000. The address `leader.telekube.local` will point to 127.0.0.1
```
Comment (a-palchikov) [inline enhancements/gep/001-loadbalancer/README.md]:
```suggestion
The domain `leader.telekube.local` will either resolve to 127.0.0.1 
```

File: enhancements/gep/001-loadbalancer/README.md [hunk 19]
```diff
@@ -0,0 +1,118 @@
...
+    type: "external|internal"
+    # gravity uses this field when type is external
+    externalAddress: "IP address | DNS address"
+```
+The names of docker images should not be changed, and they should start with `leader.telekube.local:5000/`. 
+So docker registers on master nodes by default will use port 5001 instead of 5000, 
+and loadbalancer will use port 5000. The address `leader.telekube.local` will point to 127.0.0.1 
+for the internal loadbalancer or the address of the external loadbalancer.
+
+Kubeconfig files will use a new address to point to kube-apiserver, it will be 127.0.0.1 for the internal loadbalancer or the address of the external loadbalancer.
```
Comment (a-palchikov) [inline enhancements/gep/001-loadbalancer/README.md]:
```suggestion
Kubeconfig files will either use 127.0.0.1 for the internal loadbalancer or the address of the external loadbalancer to point to the api server.
```

File: enhancements/gep/001-loadbalancer/README.md [hunk 20]
```diff
@@ -0,0 +1,118 @@
...
+
+For an external load balancer:
```
Comment (a-palchikov) [inline enhancements/gep/001-loadbalancer/README.md]:
```suggestion
For an external loadbalancer:
```
The document was consistently using this form to this point.

File: enhancements/gep/001-loadbalancer/README.md [hunk 21]
```diff
@@ -0,0 +1,118 @@
...
+The user must configure the balancer himself to open the following ports
```
Comment (a-palchikov) [inline enhancements/gep/001-loadbalancer/README.md]:
```suggestion
The user is responsible for configuring the loadbalancer to open the following ports
```

File: enhancements/gep/001-loadbalancer/README.md [hunk 22]
```diff
@@ -0,0 +1,118 @@
...
+  1. for kube-apiserver: 9443 -> (master ip addresses): 6443
+  1. for docker-registry: 5000 -> (master ip addresses): 5001
+
+For an internal load balancer:
+The software load balancer is haproxy, it will run on each node in the cluster.
+Port 9443 will be used for load balancing kube-apiserver and port 5000 will be used for load balancing in the docker registry.
```
Comment (a-palchikov) [inline enhancements/gep/001-loadbalancer/README.md]:
```suggestion
Port 9443 will be used for load-balancing kube-apiserver and port 5000 will be used for load-balancing the docker registry.
```

File: enhancements/gep/001-loadbalancer/README.md [hunk 23]
```diff
@@ -0,0 +1,118 @@
...
+The planet agent on master nodes saves and monitors the ip address in etcd using the key `/planet/cluster/${KUBE_CLUSTER_ID}/masters/${MASTER_IP}` with TTL 4 hours
+The planet agent on each node, including the master is tracing all keys
```
Comment (a-palchikov) [inline enhancements/gep/001-loadbalancer/README.md]:
```suggestion
The planet agent on each node, including master nodes will be watching all keys in `/planet/cluster/${KUBE_CLUSTER_ID}/masters/` to change the haproxy configuration and reload it if necessary.
```

File: enhancements/gep/001-loadbalancer/README.md [hunk 24]
```diff
@@ -0,0 +1,118 @@
...
+
+For an external load balancer:
+The user must configure the balancer himself to open the following ports
+  1. for kube-apiserver: 9443 -> (master ip addresses): 6443
+  1. for docker-registry: 5000 -> (master ip addresses): 5001
+
+For an internal load balancer:
+The software load balancer is haproxy, it will run on each node in the cluster.
+Port 9443 will be used for load balancing kube-apiserver and port 5000 will be used for load balancing in the docker registry.
+The planet agent on master nodes saves and monitors the ip address in etcd using the key `/planet/cluster/${KUBE_CLUSTER_ID}/masters/${MASTER_IP}` with TTL 4 hours
```
Comment (a-palchikov) [inline enhancements/gep/001-loadbalancer/README.md]:
```suggestion
The planet agent on master nodes will monitor the node IP address in etcd using the key `/planet/cluster/${KUBE_CLUSTER_ID}/masters/${MASTER_IP}` with a TTL 4 of hours.
```
Can you explain why TTL is needed?

File: enhancements/gep/001-loadbalancer/README.md (hunk 3 above)
Comment (whiterm) [reply enhancements/gep/001-loadbalancer/README.md]: I mean, if this master is restored later, it will automatically be used in the load balancer.

File: enhancements/gep/001-loadbalancer/README.md (hunk 7 above)
Comment (whiterm) [reply enhancements/gep/001-loadbalancer/README.md]: I mean adding a new node and removing an old one

File: enhancements/gep/001-loadbalancer/README.md (hunk 8 above)
Comment (whiterm) [reply enhancements/gep/001-loadbalancer/README.md:44]: They are not using the loadbalancing of the kubernetes services. They all are using the address `leader.telekube.local`

File: enhancements/gep/001-loadbalancer/README.md (hunk 4 above)
Comment (whiterm) [reply enhancements/gep/001-loadbalancer/README.md]: I don't understand your point

File: enhancements/gep/001-loadbalancer/README.md (hunk 15 above)
Comment (whiterm) [reply enhancements/gep/001-loadbalancer/README.md]: I thought it was clear if they are using leader.telekube.local to connect to the kubernetes api then it would be better to use 127.0.0.1. If for some incomprehensible purpose, then it is better to find another solution

File: enhancements/gep/001-loadbalancer/README.md (hunk 24 above)
Comment (whiterm) [reply enhancements/gep/001-loadbalancer/README.md]: I've added

File: enhancements/gep/001-loadbalancer/README.md (hunk 4 above)
Comment (a-palchikov) [reply enhancements/gep/001-loadbalancer/README.md]: Depending on what you meant by `the update`, I took the liberty of interpreting this in the context of a cluster - i.e. `cluster update` and hinted at the difference between an update and an upgrade (favoring the latter for cluster upgrades). Let me know if this is still unclear.

File: enhancements/gep/001-loadbalancer/README.md (hunk 8 above)
Comment (a-palchikov) [reply enhancements/gep/001-loadbalancer/README.md:44]: Right.

File: enhancements/gep/001-loadbalancer/README.md (hunk 7 above)
Comment (a-palchikov) [reply enhancements/gep/001-loadbalancer/README.md]: I would rephrase as:
```
When adding/removing master nodes, the agent on each node should automatically reflect this in the local loadbalancer configuration.
```
